presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1570,6 +1571,7 @@ private Scope computeAndAssignOrderByScope(OrderBy node, Scope sourceScope, Scop
             // ORDER BY should "see" both output and FROM fields during initial analysis and non-aggregation query planning
             Scope orderByScope = Scope.builder()
                     .withParent(sourceScope)
+                    .withScopeId(outputScope.getScopeId())
```

Comment:
I don't think @martint meant that the concept of `RelationId` is wrong. The concept itself is perfectly fine. However, I understand that `RelationId` shouldn't be part of `RelationType` but rather moved to Scope. This means that Scope resolves fields from given `Relation` (expressed by `RelationId`).

Scopes are not meant to uniquely identify resolved fields. They are just field (names) resolution mechanism that models hierarchy between relations. However, field still belongs to relation, not scope. As you noted, one can build different hierarchies of relations depending on resolution semantics. The same relation might be part of different scopes (e.g: ORDER BY sees both output and from fields but top QuerySpecification sees only output fields of lower relation).

Why not just move `RelationId` out of `RelationType` into `Scope`? It should work as it was before but with clearer separation between `RelationType` and `RelationId`?